chain/src/main/scala/org/bitcoins/chain/models/BlockHeaderDAO.scala

```diff
@@ -71,37 +71,29 @@ case class BlockHeaderDAO()(implicit
      * To avoid making many database reads, we make one database read for all
      * possibly useful block headers.
      */
-    val headersF = getBetweenHeights(from = height, to = child.height - 1)
+    lazy val headersF = getBetweenHeights(from = height, to = child.height - 1)
 
     /*
      * We then bucket sort these headers by height so that any ancestor can be found
      * in linear time assuming a bounded number of contentious tips.
      */
     val headersByHeight: Array[Vector[BlockHeaderDb]] =
-      new Array[Vector[BlockHeaderDb]](_length = (child.height - height).toInt)
-
-    /*
-     * I believe Array's of Objects are instantiated with null, which is evil,
-     * and so we start by giving each element of the array a Vector.empty.
-     */
-    headersByHeight.indices.foreach(index =>
-      headersByHeight(index) = Vector.empty)
+      Array.fill(child.height - height)(Vector.empty[BlockHeaderDb])
 
     // Bucket sort
     headersF.map { headers =>
       headers.foreach { header =>
-        val index = (header.height - height).toInt
+        val index = header.height - height
         headersByHeight(index) = headersByHeight(index).:+(header)
       }
 
-      // Now that the bucket sort is done, we get rid of mutability
-      val groupedByHeightHeaders: List[Vector[BlockHeaderDb]] =
-        headersByHeight.toList
+      val groupedByHeightHeaders: Array[Vector[BlockHeaderDb]] =
```

Comment:
Since the Array is not leaked via the return type of the method, there is little benefit to convert here. In terms of runtime, it takes 5x longer to convert to List and reverse, than to just reverse the array
```
@ (1 to 1000).map(_ => takes(() => a.reverse   )).sum / 1000 
res59: Long = 2659141L

@ (1 to 1000).map(_ => takes(() => a.toList.reverse   )).sum / 1000 
res60: Long = 12484410L
```

Reply:
Wow! Good to know. Thanks!